Index support collaboration

christopherjwang opened the pull request.

yzubkov committed:
Add support for managing collection indexes

yzubkov committed:
Add support for managing collection indexes

yzubkov committed:
Add support for managing collection indexes

yzubkov committed:
Replacing sprintf for strcpy

yzubkov committed:
Adding example of mixed type, compound index

yzubkov committed:
Simplifying the indexing example

yzubkov committed:
Changing the createIndex return type to string index name

yzubkov committed:
Passing options in lua table without converting to BSON

yzubkov committed:
Adding spacing

yzubkov committed:
Simplifying return value

yzubkov committed:
Switching to lua_tostring() from strcpy()

yzubkov committed:
Simplify option values

Commit:
Merge pull request #62 from yzubkov/master

Baseline for indexing support. Using this shared branch to continue working on Index support

Commit:
Use numbers in Lua 5.1/5.2 for index options

Commit:
C driver 1.2 compatability

mongoc_collection_create_index_with_opts was added after v1.2
https://github.com/mongodb/mongo-c-driver/commit/279c7fe7fbc4309d727d9762014e6aa49483d19e

christopherjwang commented:
@yzubkov check out #67 and #68

christopherjwang commented on c_wrapper/lua-mongoc-collection.c:
```diff
@@ -629,3 +629,150 @@ lua_mongo_collection_destroy(lua_State *L)
+    }
+
+    if ( lua_istable(L, options_index) ){
+        lua_getfield(L, options_index, "name"); //String
```
There's no need to append `//String` to these lines. This is only because we call 
```
if( lua_isstring(L, -9) ){
  opt.name = lua_tostring(L, -9);
}
```
right after. It self documents itself.

christopherjwang commented on c_wrapper/lua-mongoc-collection.c:
```diff
@@ -629,3 +629,150 @@ lua_mongo_collection_destroy(lua_State *L)
+        lua_getfield(L, options_index, "textIndexVersion"); //Integer
+        lua_getfield(L, options_index, "weights"); //Table
+
+        if( lua_isstring(L, -9) ){
```
as for style, I'd prefer to use 
`if(lua_isstring(L, -9)){`

christopherjwang commented on c_wrapper/lua-mongoc-collection.c:
```diff
@@ -629,3 +629,150 @@ lua_mongo_collection_destroy(lua_State *L)
+    }
+
+    collection = (collection_t *) luaL_checkudata(L, 1, "lua_mongoc_collection");
+    created=mongoc_collection_create_index_with_opts(collection->c_collection, &keys, &opt, NULL, NULL, &error);
```
needs to be `created = mongoc....`, not `created=mongoc...`

christopherjwang commented on examples/indexing.lua:
```diff
@@ -0,0 +1,45 @@
+col:insert_many(sample)
+printResult( col:find({}) )
+
+local res=col:createIndex( {name='text', txt='text'}, { 
```
no space after (

christopherjwang commented on examples/indexing.lua:
```diff
@@ -0,0 +1,45 @@
+printResult( col:find({}) )
+
+local res=col:createIndex( {name='text', txt='text'}, { 
+	name='myindex', 
```
need space before and after `=`

christopherjwang commented on mongorover/MongoCollection.lua:
```diff
@@ -265,6 +265,34 @@ local MongoCollection = {__mode="k"}
+	-- Example usage at @{indexing.lua}.
+	-- @tparam table of fields and their corresponding index types
+	-- @tparam table of options to be used when creating index
+	-- @return string index name on success
```
This function won't return nil, it will just error out because of this [line](https://github.com/mongodb-labs/mongorover/pull/66/files#diff-b3084e8d014c2b0adf0231692fda4e23R713), right? This might have to change in behavior to whatever pymongo does.

Commit:
Merge pull request #68 from christopherjwang/christopherjwang/c_driver_version_function_call

C driver 1.2 compatability

Commit:
Merge pull request #67 from christopherjwang/christopherjwang/lua_51_52

Use numbers in Lua 5.1/5.2 for index options

yzubkov committed:
Fixing documentation parameters and return value (#75)

* Fixing documentation parameters and return value

* Fix up documentation for index methods

yzubkov committed:
Updated indexing example with proper string concatenation. (#72)

* Updated indexing example with proper string concatenation.

* Minor fixes

* Fixing dropIndex() returning nil

* Fixing dropIndex() returning nil

yzubkov committed:
Returning nil value from lua_mongo_collection_drop_index() (#74)

* Returning nil value from lua_mongo_collection_drop_index()

* Adding bson_free()

Commit:
Formatting

Commit:
Correct error handling for find_indexes cursor

christopherjwang commented:
@yzubkov,

I made some changes to your code.
You want to have this type of format:
`if (something) {`
and not:
`if(something){`

It's harder to read that way. Although the rest of this codebase isn't super easy to read, that is a step in the right direction.

`error.code` isn't a valid way of checking that something is valid or not. It's actually `mongoc_cursor_error(...)` seen [here](http://mongoc.org/libmongoc/1.4.0/cursors.html#cursor-errors). That's my bad, it wasn't very apparent at first.

I'll try to get more tests out tomorrow.


A couple things for you to do if you want:

- [ ] Make me a collaborator on your fork so I can alter your PRs before merging.
- [ ] I'd like the functions to be like `create_index` and not `createIndex`.
- [ ] I'd like you to use `list_indexes` instead of `findIndexes`, as that is what pymongo [does](http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.list_indexes)
- [ ] Start implementing [index_information](http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.index_information) and [reindex](http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.reindex) and add those to your examples script that you've been working on so it's easier for me to get up to speed.
- [ ] Ensure the behavior in mongorover's index functions follow pymongo's as closely as possible.

yzubkov commented:
Thank you very much for merging the PR's into the "index_support" branch. I have updated my branch to reflect what is in "index_support" and additionally made the recommended changes to re-name the functions to have the underscore character. 

I have also renamed a parameter `textIndexVersion` to `version` ,  because it has nothing to do with text indexing and is just a plain index version as mentioned in this mogoDB documentation: [here](https://docs.mongodb.com/v3.2/release-notes/3.2-compatibility/#version-0-indexes). Should I make these same PR's into the "index_support" branch?

Because we want to stay close to PyMongo, I want to document the following difference that I noticed when we are asking C driver to auto-generate an index name based on the supplied list of keys/arguments. When we supply the following keys: `{name = 'text', cost = -1}`, the C function `opt.name = mongoc_collection_keys_to_index_string(&keys);` will return `name_text_cost_0`. It will return the same index name if we change the directional indicator from descending to ascending and give it `{name = 'text', cost = 1}`. In PyMongo the generated index name would be different between the two scenarios, as you can see [here](https://github.com/mongodb/mongo-python-driver/blob/ba6ef88d10eec5a1f7d57d12d96ac9f5fe14184d/test/test_collection.py#L246)

I am glad that you helped me get this as far as we got it currently. My focus from the start was to make sure we have the baseline indexing support so that I am able to do text searches on a collection. I do feel that the additional `index_information` and `reindex` commands will be beneficial to this driver and I would be glad to help develop some tests and modify example script to demo these new functions, although not sure how soon.

christopherjwang commented:
Yes, PR's into the `index_support` branch will work just fine.

I would just use the C driver's auto-generate functionality.

Not to worry, thanks for helping out. I'll at least take what you've started and start moving towards an actual feature.

yzubkov committed:
Renaming function find_indexes to list_indexes and other naming conventions

* Updating function name to list_indexes

Also changing "textIndexVersion" to plain "version" parameter name.

* Updating function name to list_indexes

Also changing "textIndexVersion" to plain "version" parameter name.

* Updating function name to list_indexes

* Updating function name to list_indexes

christopherjwang commented:
Thanks for working with me @yzubkov, I'm going to close this PR because I don't think I'll be able to provide the necessary support + debugging of a release. Sorry I didn't close this PR and archive this project earlier.
